File: vertex-cache-server/src/main/java/com/vertexcache/core/command/impl/GetSecondaryIdxOneCommand.java
/*
 * Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.vertexcache.core.command.impl;

import com.vertexcache.common.log.LogHelper;
import com.vertexcache.common.util.StringUtil;
import com.vertexcache.core.cache.CacheAccessService;
import com.vertexcache.core.command.BaseCommand;
import com.vertexcache.core.command.CommandResponse;
import com.vertexcache.core.command.argument.ArgumentParser;
import com.vertexcache.core.validation.validators.KeyValidator;
import com.vertexcache.module.restapi.model.ApiParameter;
import com.vertexcache.server.session.ClientSessionContext;

/**
 * Command used to retrieve all cache entries that match a given secondary index (idx1).
 *
 * Returns a collection of values whose primary keys are associated with the specified idx1 value.
 * Useful for grouped lookups or category-based access patterns.
 *
 * Does not support filtering on tertiary index (idx2); use GetTertiaryIdxTwoCommand for that.
 *
 * Requires READ or higher privileges to execute.
 */
public class GetSecondaryIdxOneCommand extends BaseCommand<String> {

    public static final String COMMAND_KEY = "GETIDX1";

    @Override
    protected String getCommandKey() {
        return COMMAND_KEY;
    }

    @Override
    public CommandResponse execute(ArgumentParser argumentParser, ClientSessionContext session) {
        CommandResponse response = new CommandResponse();

        try {
            if (argumentParser.getPrimaryArgument().getArgs().size() != 1) {
                response.setResponseError("GETIDX1 command requires a single argument: the secondary index (IDX1) key.");
                return response;
            }

            String idxKey = argumentParser.getPrimaryArgument().getArgs().getFirst();

            try {
                new KeyValidator(ApiParameter.IDX1.value(), idxKey).validate();
            } catch (Exception ex) {
                response.setResponseError(ex.getMessage());
                return response;
            }

            CacheAccessService service = new CacheAccessService();
            String value = service.getBySecondaryIdx1(session, idxKey);

            if (value != null) {
                response.setResponse(StringUtil.esacpeQuote(value));
            } else {
                response.setResponseNil();
            }

        } catch (Exception ex) {
            response.setResponseError("GETIDX1 command failed. Check logs.");
            LogHelper.getInstance().logFatal("[GetSecondaryIdxOneCommand] error: " + ex.getMessage(), ex);
        }

        return response;
    }
}

File: client-sdks/elixir/sdk/command/impl/get_secondary_idx_one_command.ex
# ------------------------------------------------------------------------------
# Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------

defmodule VertexCacheSdk.Command.Impl.GetSecondaryIdxOneCommand do
  @moduledoc """
  Handles the GET Secondary Idx (idx1) command in VertexCache.

  Retrieves the value for a given key from the cache using a secondary index.
  Returns an error if the key is missing or expired.

  Requires the client to have READ, READ_WRITE, or ADMIN access.
  """

  @behaviour VertexCacheSdk.Command

  alias VertexCacheSdk.Model.{CommandResult, GetResult, VertexCacheSdkException}
  alias VertexCacheSdk.Comm.ClientConnector
  alias VertexCacheSdk.Command.CommandBase

  @type t :: %__MODULE__{key: String.t()}
  defstruct key: nil

  @spec new(String.t()) :: t
  def new(key) when is_binary(key) do
    if String.trim(key) == "" do
      raise VertexCacheSdkException, message: "GET By Secondary Index (idx1) command requires a non-empty key"
    else
      %__MODULE__{key: key}
    end
  end

  def new(_),
      do: raise VertexCacheSdkException,
          message: "GET By Secondary Index (idx1) command requires a non-empty key"

  @spec execute(t, ClientConnector.t()) :: GetResult.t()
  def execute(%__MODULE__{key: key}, connector) do
    CommandBase.execute("GETIDX1 " <> key, connector, &parse_response/1)
    |> to_get_result()
  end

  @spec parse_response(String.t()) :: String.t() | no_return()
  def parse_response("(nil)"), do: "(nil)"

  def parse_response(response) when is_binary(response) do
    if String.starts_with?(response, "ERR") do
      raise "GETIDX1 failed: #{response}"
    else
      response
    end
  end

  @spec to_get_result(CommandResult.t()) :: GetResult.t()
  defp to_get_result(%CommandResult{success: true, status_message: "(nil)"} = result) do
    %GetResult{success: true, status_message: result.status_message, value: nil}
  end

  defp to_get_result(%CommandResult{success: true, status_message: value}) do
    %GetResult{success: true, status_message: value, value: value}
  end

  defp to_get_result(%CommandResult{success: false, status_message: error}) do
    %GetResult{success: false, status_message: error, value: nil}
  end
end
